File: Android/EasyBudget/app/src/main/java/com/benoitletondor/easybudgetapp/view/welcome/Onboarding3Fragment.java
package com.benoitletondor.easybudgetapp.view.welcome;


import android.os.Bundle;
import android.app.Fragment;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import com.benoitletondor.easybudgetapp.R;
import com.benoitletondor.easybudgetapp.helper.CurrencyHelper;
import com.benoitletondor.easybudgetapp.model.Expense;
import com.benoitletondor.easybudgetapp.model.db.DB;

import java.util.Date;

/**
 * Onboarding step 3 fragment
 *
 * @author Benoit LETONDOR
 */
public class Onboarding3Fragment extends OnboardingFragment
{
    private TextView moneyTextView;
    private EditText amountEditText;
    private Button nextButton;

// -------------------------------------->

    /**
     * Required empty public constructor
     */
    public Onboarding3Fragment()
    {

    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
    {
        // Inflate the layout for this fragment
        View v = inflater.inflate(R.layout.fragment_onboarding3, container, false);

        DB db = getDB();

        int amount = 0;
        if( db != null )
        {
            amount = db.getBalanceForDay(new Date());
        }

        moneyTextView = (TextView) v.findViewById(R.id.onboarding_screen3_initial_amount_money_tv);
        setCurrency();

        amountEditText = (EditText) v.findViewById(R.id.onboarding_screen3_initial_amount_et);
        amountEditText.setText(String.valueOf(amount));
        amountEditText.addTextChangedListener(new TextWatcher()
        {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after)
            {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count)
            {

            }

            @Override
            public void afterTextChanged(Editable s)
            {
                setButtonText();
            }
        });

        nextButton = (Button) v.findViewById(R.id.onboarding_screen3_next_button);
        nextButton.setOnClickListener(new View.OnClickListener()
        {
            @Override
            public void onClick(View v)
            {
                DB db = getDB();
                if (db != null)
                {
                    int currentBalance = db.getBalanceForDay(new Date());
                    int newBalance = Integer.valueOf(amountEditText.getText().toString());

                    if (newBalance == currentBalance)
                    {
                        // Nothing to do, balance hasn't change
                        return;
                    }

                    int diff = newBalance - currentBalance;

                    final Expense expense = new Expense(getResources().getString(R.string.adjust_balance_expense_title), -diff, new Date());
                    db.persistExpense(expense);
                }

                next(v);
            }
        });
        setButtonText();

        return v;
    }

    @Override
    public void setUserVisibleHint(boolean isVisibleToUser)
    {
        super.setUserVisibleHint(isVisibleToUser);

        if( isVisibleToUser ) // Update values on display
        {
            setCurrency();
            setButtonText();
        }
    }

    @Override
    public int getStatusBarColor()
    {
        return R.color.onboarding_3_statusbar;
    }

// -------------------------------------->

    private void setCurrency()
    {
        if( moneyTextView != null ) // Will be null if view is not yet created
        {
            moneyTextView.setText(CurrencyHelper.getUserCurrency(getActivity()).getSymbol());
        }
    }

    private void setButtonText()
    {
        if( nextButton != null )
        {
            String valueString = amountEditText.getText().toString();
            int value = "".equals(valueString) ? 0 : Integer.valueOf(valueString);

            nextButton.setText(getActivity().getString(R.string.onboarding_start_with_amount, CurrencyHelper.getFormattedCurrencyString(getActivity(), value)));
        }
    }
}

File: Android/EasyBudget/app/src/main/java/com/benoitletondor/easybudgetapp/view/welcome/Onboarding3Fragment.kt
/*
 *   Copyright 2024 Benoit LETONDOR
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */

package com.benoitletondor.easybudgetapp.view.welcome

import android.content.Context
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import com.benoitletondor.easybudgetapp.R
import com.benoitletondor.easybudgetapp.databinding.FragmentOnboarding3Binding
import com.benoitletondor.easybudgetapp.db.DB
import com.benoitletondor.easybudgetapp.helper.*
import com.benoitletondor.easybudgetapp.model.Expense
import com.benoitletondor.easybudgetapp.parameters.Parameters
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.*
import java.time.LocalDate
import javax.inject.Inject

/**
 * Onboarding step 3 fragment
 *
 * @author Benoit LETONDOR
 */
@AndroidEntryPoint
class Onboarding3Fragment : OnboardingFragment<FragmentOnboarding3Binding>() {
    @Inject lateinit var parameters: Parameters
    @Inject lateinit var db: DB

    override val statusBarColor: Int
        get() = R.color.secondary_dark

    private val amountValue: Double
        get() {
            val valueString = binding?.onboardingScreen3InitialAmountEt?.text.toString()

            return try {
                if ( "" == valueString || "-" == valueString) 0.0 else java.lang.Double.valueOf(valueString)
            } catch (e: Exception) {
                val context = context ?: return 0.0

                MaterialAlertDialogBuilder(context)
                    .setTitle(R.string.adjust_balance_error_title)
                    .setMessage(R.string.adjust_balance_error_message)
                    .setNegativeButton(R.string.ok) { dialog, _ -> dialog.dismiss() }
                    .show()

                Logger.warning("An error occurred during initial amount parsing: $valueString", e)
                return 0.0
            }

        }

    override fun onCreateBinding(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): FragmentOnboarding3Binding = FragmentOnboarding3Binding.inflate(inflater, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        viewLifecycleScope.launch {
            val amount = withContext(Dispatchers.Default) {
                -db.getBalanceForDay(LocalDate.now())
            }

            binding?.onboardingScreen3InitialAmountEt?.setText(if (amount == 0.0) "0" else amount.toString())
        }

        setCurrency()

        binding?.onboardingScreen3InitialAmountEt?.preventUnsupportedInputForDecimals()
        binding?.onboardingScreen3InitialAmountEt?.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {

            }

            override fun afterTextChanged(s: Editable) {
                setButtonText()
            }
        })

        binding?.onboardingScreen3NextButton?.setOnClickListener { button ->
            viewLifecycleScope.launch {
                withContext(Dispatchers.Default) {
                    val currentBalance = -db.getBalanceForDay(LocalDate.now())
                    val newBalance = amountValue

                    if (newBalance != currentBalance) {
                        val diff = newBalance - currentBalance

                        val expense = Expense(resources.getString(R.string.adjust_balance_expense_title), -diff, LocalDate.now(), true)
                        db.persistExpense(expense)
                    }
                }

                // Hide keyboard
                try {
                    val imm = context?.getSystemService(Context.INPUT_METHOD_SERVICE) as? InputMethodManager
                    imm?.hideSoftInputFromWindow(binding?.onboardingScreen3InitialAmountEt?.windowToken, 0)
                } catch (e: Exception) {
                    Logger.error("Error while hiding keyboard", e)
                }

                next(button)
            }
        }

        setButtonText()
    }

    override fun onResume() {
        super.onResume()

        setCurrency()
        setButtonText()
    }

// -------------------------------------->

    private fun setCurrency() {
        binding?.onboardingScreen3InitialAmountMoneyTv?.text = parameters.getUserCurrency().symbol
    }

    private fun setButtonText() {
        binding?.onboardingScreen3NextButton?.text = getString(R.string.onboarding_screen_3_cta, CurrencyHelper.getFormattedCurrencyString(parameters, amountValue))
    }
}
